### ISSUE TITLE
alcSetError when opening capture device for stream 

### ISSUE THREAD

**Issue by TLMcNulty** [kcat/openal-soft#389]
I'm trying to use a bluetooth helmet communicator with some software that uses openal for it's interactions with ALSA.

Receiving audio works well and for a while the other day I was able to transmit but the quality was very low. I've been trying to fix it but now I can't transmit anything at all. 

When the device is opened an error is generated. 

```
Transmit start...
AL lib: (WW) alcSetError: Error generated on device 0xa2d00fb0, code 0xa004
AL lib: (WW) alcSetError: Error generated on device 0xa2d00fb0, code 0xa004
AL lib: (WW) alcSetError: Error generated on device 0xa2d00fb0, code 0xa004
AL lib: (WW) alcSetError: Error generated on device 0xa2d00fb0, code 0xa004
```

Can you help decode that error code? Is it a missing device?

**Comment by kcat** [comment]
`0xa004` is `ALC_INVALID_VALUE`. What calls do you make that cause the error?

**Comment by TLMcNulty** [comment]
`CaptureStart`

**Comment by TLMcNulty** [comment]
All the other calls to this source seem to be working correctly but when starting a stream it errors and no sound is collected. 

```
AL lib: (II) alcCaptureOpenDevice: Capture format: Mono, Signed Short, 16000hz, 960 update size x1
AL lib: (II) GetProcBinary: Got filename: filename
AL lib: (II) GetConfigValue: Key pulse/allow-moves not found
AL lib: (II) ALCpulseCapture_open: Connecting to "(default)"
AL lib: (II) alcCaptureOpenDevice: Created device 0x1d2bc20, "Monitor of BT8"
```

**Comment by kcat** [comment]
`alcCaptureStart` can't set an `ALC_INVALID_VALUE` error. It would have to be some other call with the device that generates the error, either before `alcCaptureStart`, or after it in between the `alcGetError`.

**Comment by TLMcNulty** [comment]
Understood. I saw in the wiki it's set by functions related to contexts? When would a context come in the process of opening a device?

Creating a device, opening, creating context:

```
((creating a stream object in go))
		s.deviceSource = openal.CaptureOpenDevice("", gumble.AudioSampleRate, openal.FormatMono16, uint32(s.sourceFrameSize)) #Capturing device results in source.

	s.deviceSink = openal.OpenDevice("") s
	s.contextSink = s.deviceSink.CreateContext()
	s.contextSink.Activate()

	s.link = client.Config.AttachAudio(s)

	return s, nil
```

Logs of above actions:

```
AL lib: (II) alcCaptureOpenDevice: Capture format: Mono, Signed Short, 16000hz, 960 update size x1
AL lib: (II) GetProcBinary: Got filename: talkiepi2
AL lib: (II) GetConfigValue: Key pulse/allow-moves not found
AL lib: (II) ALCpulseCapture_open: Connecting to "(default)"
AL lib: (II) alcCaptureOpenDevice: Created device 0x1d2bc20, "Monitor of BT8"
*AL lib: (II) GetConfigValue: Key pulse/allow-moves not found*
*AL lib: (II) GetConfigValue: Found pulse/allow-moves = "yes"*
AL lib: (II) ALCpulsePlayback_open: Connecting to "(default)"

AL lib: (II) alcOpenDevice: Created device 0x1d346a0, "BT8"
AL lib: (II) UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x3
AL lib: (II) ALCpulsePlayback_sinkInfoCallback: Active port: headset-output (Headset)

AL lib: (II) ALCpulsePlayback_bufferAttrCallback: minreq=8192, tlength=32376, prebuf=0

Channel 'Root' has 2 participants
AL lib: (II) alcCreateContext: Created context 0x1d2ad50
Transmit start...
AL lib: (WW) alcSetError: Error generated on device 0x1d2bc20, code 0xa004
AL lib: (WW) alcSetError: Error generated on device 0x1d2bc20, code 0xa004

```

Transmitting:

```
AL lib: (II) alcCaptureOpenDevice: Capture format: Mono, Signed Short, 16000hz, 960 update size x1
AL lib: (II) GetProcBinary: Got filename: talkiepi2
AL lib: (II) GetConfigValue: Found pulse/allow-moves = "yes"
AL lib: (II) ALCpulseCapture_open: Connecting to "(default)"
AL lib: (II) alcCaptureOpenDevice: Created device 0xa43035d0, "Monitor of Sena 20S v1.7.7"
Transmit start...
AL lib: (WW) alcSetError: Error generated on device 0xa43035d0, code 0xa004
AL lib: (WW) alcSetError: Error generated on device 0xa43035d0, code 0xa004
AL lib: (WW) alcSetError: Error generated on device 0xa43035d0, code 0xa004
AL lib: (WW) alcSetError: Error generated on device 0xa43035d0, code 0xa004
Transmit stop...
```

As far as I can tell the error is only generated during the transmit loop. I will investigate there.

**Comment by TLMcNulty** [comment]
Could CaptureSamples generate that error?

```
			buff := s.deviceSource.CaptureSamples(uint32(frameSize))
			if len(buff) != frameSize*2 {
				continue
			}
```

Could my frameSize be incorrect?

**Comment by kcat** [comment]
> Could CaptureSamples generate that error?

Yes, if you're asking it to capture more samples than are available, it's an invalid value. You have to wait until the device has enough samples before capturing them, or you can just capture however many samples it has at the time. You can check how many samples the device currently has by querying it, though I don't know enough about that language wrapper to say how to do it (in C/C++, it's `alcGetIntegerv(device, ALC_CAPTURE_SAMPLES, 1, &count);`).

**Comment by TLMcNulty** [comment]
That's very helpful, thank you. I will continue testing.

**Comment by TLMcNulty** [comment]
I resolved my issues. Wrapping up this thread. 

**To begin with use case**
I'm building a headless Raspberry Pi that is installed in a car that participates in the 24 Hours of Lemons. This Pi runs our communications as well as a simple OBD logger that pushes data to an influxdb instance. 

I'm using cheap bluetooth helmet communicators from ebay, and with my Raspberry Pi 4 there are a few gotchyas. 

- Make sure to set headset as the active profile for the device.
- Disable time scheduling and use pulseaudio interrupts.
- Route SCO packets over HCI so the information actually makes it way over.
- Ensure system configurations are expecting the different packet size HSP results in
- Ensure software configurations are expecting the different packet size HSP results in.

[Additional docs and commands](https://docs.wotlemons.com/doku.php?id=r_d:comms)

I changed my default packet size to 80. That, plus the rest of the configuration got me interacting with the device as I'd expect. 

Many thanks.
